frame/assets/src/tests.rs

```diff
@@ -705,19 +705,115 @@ fn set_team_should_work() {
 	new_test_ext().execute_with(|| {
 		assert_ok!(Assets::force_create(RuntimeOrigin::root(), 0, 1, true, 1));
 		assert_ok!(Assets::mint(RuntimeOrigin::signed(1), 0, 1, 100));
 		assert_ok!(Assets::mint(RuntimeOrigin::signed(1), 0, 2, 100));
 		assert_eq!(Assets::balance(0, 1), 100);
 		assert_eq!(Assets::balance(0, 2), 100);
 		assert_ok!(Assets::freeze(RuntimeOrigin::signed(1), 0, 2));
+		// can transfer to `2`
 		assert_ok!(Assets::transfer(RuntimeOrigin::signed(1), 0, 2, 50));
+		// cannot transfer from `2`
+		assert_noop!(Assets::transfer(RuntimeOrigin::signed(2), 0, 1, 25), Error::<Test>::Frozen);
+		assert_eq!(Assets::balance(0, 1), 50);
 		assert_eq!(Assets::balance(0, 2), 150);
 	});
 }
 
+#[test]
+fn touching_and_freezing_account_with_zero_asset_balance_should_work() {
+	new_test_ext().execute_with(|| {
+		// need some deposit for the touch
+		Balances::make_free_balance_be(&2, 100);
+		assert_ok!(Assets::force_create(RuntimeOrigin::root(), 0, 1, true, 1));
+		assert_ok!(Assets::mint(RuntimeOrigin::signed(1), 0, 1, 100));
+		assert_eq!(Assets::balance(0, 1), 100);
+		assert_eq!(Assets::balance(0, 2), 0);
+		// cannot freeze an account that doesn't have an `Assets` entry
+		assert_noop!(Assets::freeze(RuntimeOrigin::signed(1), 0, 2), Error::<Test>::NoAccount);
+		assert_ok!(Assets::touch(RuntimeOrigin::signed(2), 0));
+		// now it can be frozen
+		assert_ok!(Assets::freeze(RuntimeOrigin::signed(1), 0, 2));
+		// can transfer to `2` even though its frozen
+		assert_ok!(Assets::transfer(RuntimeOrigin::signed(1), 0, 2, 50));
+		// cannot transfer from `2`
+		assert_noop!(Assets::transfer(RuntimeOrigin::signed(2), 0, 1, 25), Error::<Test>::Frozen);
+		assert_eq!(Assets::balance(0, 1), 50);
+		assert_eq!(Assets::balance(0, 2), 50);
+	});
+}
+
+#[test]
+fn freeze_creating_works() {
+	new_test_ext().execute_with(|| {
+		Balances::make_free_balance_be(&1, 100);
+		assert_ok!(Assets::force_create(RuntimeOrigin::root(), 0, 1, true, 1));
+		assert_ok!(Assets::mint(RuntimeOrigin::signed(1), 0, 1, 100));
+		assert_eq!(Assets::balance(0, 1), 100);
+		assert_eq!(Assets::balance(0, 2), 0);
+		// cannot freeze an account that doesn't have an `Assets` entry
+		assert_noop!(Assets::freeze(RuntimeOrigin::signed(1), 0, 2), Error::<Test>::NoAccount);
+		// can freeze the account by placing a deposit
+		assert_ok!(Assets::freeze_creating(RuntimeOrigin::signed(1), 0, 2));
+		// `1` had to reserve `AssetAccountDeposit` to create `Account(id, 2)`
+		assert_eq!(Balances::reserved_balance(&1), 10);
+		// can transfer to `2` even though it's frozen
+		assert_ok!(Assets::transfer(RuntimeOrigin::signed(1), 0, 2, 50),);
+		// cannot transfer from `2`
+		assert_noop!(Assets::transfer(RuntimeOrigin::signed(2), 0, 1, 25), Error::<Test>::Frozen);
+		// asset balances unchanged
+		assert_eq!(Assets::balance(0, 1), 50);
+		assert_eq!(Assets::balance(0, 2), 50);
+		// refund goes back to `1`
+		assert_ok!(Assets::thaw(RuntimeOrigin::signed(1), 0, 2));
```

Comment:
Is it necessary?

Reply:
Yes because the point of the `assert_noop` below is that the Freezer cannot burn the 50 units on `2`. We know `refund_other` would fail with `Frozen`, not `WouldBurn`, without this call.